proxy/util/gcloudutil.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+	"golang.org/x/oauth2"
+)
+
+// GcloudConfigData represents the data returned by `gcloud config config-helper`.
+type GcloudConfigData struct {
+	Configuration struct {
+		Properties struct {
+			Core struct {
+				Project string
+				Account string
+			}
+		}
+	}
+	Credential struct {
+		AccessToken string    `json:"access_token"`
+		TokenExpiry time.Time `json:"token_expiry"`
+	}
+}
+
+func (cfg *GcloudConfigData) oauthToken() *oauth2.Token {
+	return &oauth2.Token{
+		AccessToken: cfg.Credential.AccessToken,
+		Expiry:      cfg.Credential.TokenExpiry,
+	}
+}
+
+type GcloudStatusCode int
+
+const (
+	GcloudOk GcloudStatusCode = iota
+	GcloudNotFound
+	// generic execution failure error not specified above.
+	GcloudExecErr
+)
+
+type GcloudError struct {
+	GcloudError error
+	Status      GcloudStatusCode
+}
+
+func (e *GcloudError) Error() string {
+	return e.GcloudError.Error()
+}
+
+// GcloudConfig returns a GcloudConfigData object or an error of type *GcloudError.
+func GcloudConfig() (*GcloudConfigData, error) {
+	gcloudCmd := "gcloud"
+	if runtime.GOOS == "windows" {
+		gcloudCmd = gcloudCmd + ".cmd"
+	}
+
+	if _, err := exec.LookPath(gcloudCmd); err != nil {
+		return nil, &GcloudError{err, GcloudNotFound}
+	}
+
+	buf := new(bytes.Buffer)
+	cmd := exec.Command(gcloudCmd, "--format", "json", "config", "config-helper")
+	cmd.Stdout = buf
+
+	if err := cmd.Run(); err != nil {
+		if strings.Contains(err.Error(), "executable file not found") {
+			return nil, &GcloudError{err, GcloudNotFound}
+		}
+		logging.Errorf("Error detecting gcloud project: %v", err)
+		return nil, &GcloudError{err, GcloudExecErr}
+	}
+
+	data := &GcloudConfigData{}
+
+	if err := json.Unmarshal(buf.Bytes(), data); err != nil {
+		logging.Errorf("Failed to unmarshal bytes from gcloud: %v", err)
```

Comment:
There's a tott about logging errors or returning them, not both. It's called like "contextful logs", and suggests annotating errors you return rather than  printing and returning errors.

Reply:
Yeah, agreed.

In this case this function is an extraction of the function that invoked gcloud before, so I left the logging/return error behavior as is.